osgp/platform/osgp-secret-management/src/test/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementServiceTest.java

```diff
@@ -177,28 +180,128 @@ public void storeSecrets() throws Exception {
+    // WHEN
+    when(this.secretRepository.getSecretCount(
+            SOME_DEVICE, SecretType.E_METER_MASTER_KEY, SecretStatus.NEW))
+        .thenReturn(1);
+    when(this.secretRepository.findSecrets(
+            SOME_DEVICE, SecretType.E_METER_MASTER_KEY, SecretStatus.NEW))
+        .thenReturn(Arrays.asList(secret));
+
+    final TypedSecret typedSecret = new TypedSecret(new byte[16], SecretType.E_METER_MASTER_KEY);
+    this.service.storeOrResetNewSecrets(SOME_DEVICE, Arrays.asList(typedSecret));
+
+    // THEN
+    final List<DbEncryptedSecret> foundSecrets =
+        this.secretRepository.findSecrets(
+            SOME_DEVICE, SecretType.E_METER_MASTER_KEY, SecretStatus.NEW);
+    assertThat(foundSecrets).hasSize(1);
+
+    verify(this.secretRepository, never()).saveAll(Arrays.asList(secret));
+    assertThat(secret.getCreationTime().getTime() - new Date().getTime()).isLessThan(10L);
+    assertThat(secret.getSecretStatus()).isEqualTo(SecretStatus.NEW);
+  }
+
+  @Test
+  public void storeNewSecretsWhenOneRecentAndOlderNewSecretAlreadyExists() throws Exception {
+    // GIVEN
+    final DbEncryptedSecret secretOldMaster = this.getNewMasterSecret(100);
+    final DbEncryptedSecret secretOlderMaster = this.getNewMasterSecret(1000);
+    final DbEncryptedSecret secretOldAuthen = this.getNewAuthenticationSecret(100);
+    final DbEncryptedSecret secretOlderAuthen = this.getNewAuthenticationSecret(1000);
+    final Date olderCreationTime = secretOlderMaster.getCreationTime();
+
+    // WHEN
+    when(this.secretRepository.getSecretCount(
+            SOME_DEVICE, SecretType.E_METER_MASTER_KEY, SecretStatus.NEW))
+        .thenReturn(2);
+    when(this.secretRepository.findSecrets(
+            SOME_DEVICE, SecretType.E_METER_MASTER_KEY, SecretStatus.NEW))
+        .thenReturn(Arrays.asList(secretOldMaster, secretOlderMaster));
+    when(this.secretRepository.getSecretCount(
+            SOME_DEVICE, SecretType.E_METER_AUTHENTICATION_KEY, SecretStatus.NEW))
+        .thenReturn(2);
+    when(this.secretRepository.findSecrets(
+            SOME_DEVICE, SecretType.E_METER_AUTHENTICATION_KEY, SecretStatus.NEW))
+        .thenReturn(Arrays.asList(secretOldAuthen, secretOlderAuthen));
+
+    final TypedSecret typedSecretMaster =
+        new TypedSecret(new byte[16], SecretType.E_METER_MASTER_KEY);
+    final TypedSecret typedSecretAuthen =
+        new TypedSecret(new byte[16], SecretType.E_METER_AUTHENTICATION_KEY);
+    this.service.storeOrResetNewSecrets(
+        SOME_DEVICE, Arrays.asList(typedSecretMaster, typedSecretAuthen));
+
+    // THEN
+    verify(this.secretRepository, never()).saveAll(Arrays.asList(secretOldMaster));
+    verify(this.secretRepository, never()).saveAll(Arrays.asList(secretOldAuthen));
+    verify(this.secretRepository, never()).saveAll(Arrays.asList(secretOldMaster, secretOldAuthen));
+
+    assertThat(secretOldMaster.getCreationTime().getTime() - new Date().getTime()).isLessThan(10L);
```

Comment:
Does this do what you would expect?
It looks like an attempt to check that the creation that was older is updated to something within the last 10 milliseconds.
Isn't `new Date().getTime()` always the bigger value than the creation time from the `DbEncryptedSecret`? That would lead to negative numbers, which are always less than `10`.

AssertJ has nicer ways to assert on time based values, that may even make the code more readable, and the exception messages more readily understandable (for instance `assertThat(secretOldMaster.getCreationTime()).isCloseTo(new Date(), 10);`).

Reply:
That is a nice assertion I didn't know of.

Reply:
Implemented three times